lib/device.ex

```diff
@@ -0,0 +1,198 @@
+                :media_service_path,
+                time_diff_from_system_secs: 0,
+                port: 80,
+                supports_media2?: false,
+                device_service_path: "/onvif/device_service"
+              ]
+
+  @spec init(Onvif.Discovery.Probe.t(), String.t(), String.t(), boolean, boolean) ::
+          {:error, any()}
+          | %Device{}
+  def init(
+        %Probe{} = probe_match,
+        username,
+        password,
+        prefer_https? \\ false,
+        prefer_ipv6? \\ false
+      ) do
+    with {:ok, device} <-
+           device_from_probe_match(probe_match, username, password, prefer_https?, prefer_ipv6?),
+         {:ok, device_with_datetime} <- get_date_time(device),
+         {:ok, updated_device} <- guess_auth(device_with_datetime) do
+      updated_device
+      |> get_services()
+      |> set_media_service_path()
+    end
+  end
+
+  defp device_from_probe_match(
+         %Probe{} = probe_match,
+         username,
+         password,
+         prefer_https?,
+         prefer_ipv6?
+       ) do
+    uri =
+      probe_match.address
+      |> get_address(prefer_ipv6?, prefer_https?)
+      |> URI.parse()
+      |> Map.put(:userinfo, "#{username}:#{password}")
+
+    device = %Device{
+      username: username,
+      password: password,
+      address: URI.to_string(uri),
+      port: uri.port,
+      scopes: probe_match.scopes
+    }
+
+    {:ok, device}
+  end
+
+  defp get_address(addresses, prefer_ipv6?, prefer_https?) do
+    # Precedence for the ipv6 and https follow the order below
+    # ipv6 and https
+    # ipv6 and http
+    # ipv4 and https
+    # ipv4 and http
+    ipv6_addresses =
+      Enum.filter(addresses, fn address ->
+        result =
+          address
+          |> URI.parse()
+          |> then(& &1.host)
+          |> :binary.bin_to_list()
+          |> :inet.parse_ipv6strict_address()
+
+        case result do
+          {:ok, _} ->
+            cond do
+              prefer_https? -> String.contains?(address, "https://")
+              true -> true
+            end
+
+          {:error, _} ->
+            false
+        end
+      end)
+
+    ipv4_addresses =
+      Enum.filter(addresses, fn address ->
+        result =
+          address
+          |> URI.parse()
+          |> then(& &1.host)
+          |> :binary.bin_to_list()
+          |> :inet.parse_ipv4strict_address()
+
+        case result do
+          {:ok, _} ->
+            cond do
+              prefer_https? -> String.contains?(address, "https://")
+              true -> true
+            end
+
+          {:error, _} ->
+            false
+        end
+      end)
+
+    cond do
+      prefer_ipv6? and Enum.count(ipv6_addresses) != 0 -> Enum.at(ipv6_addresses, 0)
+      Enum.count(ipv4_addresses) != 0 -> Enum.at(ipv4_addresses, 0)
+      true -> Enum.at(addresses, 0)
+    end
+  end
+
+  defp get_date_time(device) do
+    with {:ok, res} <- Onvif.Devices.GetSystemDateAndTime.request(device.address) do
```

Comment:
this fails because it double appends the path. When I remove the path subsequent requests fail because it expects the path in the guess auth get_device_information. Feels like we are trying to do too much in this one PR, we need to have 1. PR with the new device struct, 2. update the requests to accept a device struct instead of a URL and remove the need for a path in the device and media API contexts, 3. do the init device logic with the new Device struct requests.